src/clients/c/samples/main.c

```diff
@@ -0,0 +1,435 @@
+    ctx->size = size;
+    ctx->completed = true;
+    
+    pthread_cond_signal(&ctx->cv);
+    pthread_mutex_unlock(&ctx->lock);
+}
+
+void send_request(
+    tb_client_t client,
+    tb_packet_list_t *packets,
+    completion_context_t *ctx
+) {
+    // Locks the mutex:
+    if (pthread_mutex_lock(&ctx->lock) != 0) {
+        printf("Failed to lock mutex\n");
+        exit(-1);
+    }
+
+    // Submits the request asynchronously:
+    ctx->completed = false;
+    tb_client_submit(client, packets);
+
+    // Uses a condvar to sync this thread with the callback:
+    while (!ctx->completed) {
+        if (pthread_cond_wait(&ctx->cv, &ctx->lock) != 0) {
+            printf("Failed to wait condvar\n");
+            exit(-1);
+        }
+    }
+    
+    if (pthread_mutex_unlock(&ctx->lock) != 0) {
+        printf("Failed to unlock mutex\n");
+        exit(-1);
+    }
+}
+
+void completion_context_init(completion_context_t *ctx) {
+    if (pthread_mutex_init(&ctx->lock, NULL) != 0) {
+        printf("Failed to initialize mutex\n");
+        exit(-1);
+    }
+
+    if (pthread_cond_init(&ctx->cv, NULL) != 0) {
+        printf("Failed to initialize condition var\n");
+        exit(-1);
+    }
+}
+
+void completion_context_destroy(completion_context_t *ctx) {
+    pthread_cond_destroy(&ctx->cv);
+    pthread_mutex_destroy(&ctx->lock);
+}
+
+long long get_time_ms(void) {
+    struct timeval tv;
+    if (gettimeofday(&tv,NULL) != 0) {
+        printf("Failed to get time of day\n");
+        exit(-1);
+    }
+    return (((long long)tv.tv_sec)*1000)+(tv.tv_usec/1000);
+}
+
+#else
+
+void on_completion(
+    uintptr_t context, 
+    tb_client_t client, 
+    tb_packet_t *packet, 
+    const uint8_t *data, 
+    uint32_t size
+) {    
+    // The user_data gives context to a request:
+    completion_context_t* ctx = (completion_context_t*)packet->user_data;
+
+    // Signaling the main thread we received the reply:
+    EnterCriticalSection(&ctx->lock);
+    
+    memcpy (ctx->reply, data, size);
+    ctx->size = size;
+    ctx->completed = true;
+    
+    WakeConditionVariable(&ctx->cv);
+    LeaveCriticalSection(&ctx->lock);
+}
+
+void send_request(
+    tb_client_t client,
+    tb_packet_list_t *packets,
+    completion_context_t *ctx
+) {
+    // Locks the mutex:
+    EnterCriticalSection(&ctx->lock);
+
+    // Submits the request asynchronously:
+    ctx->completed = false;
+    tb_client_submit(client, packets);
+
+    // Uses a condvar to sync this thread with the callback:
+    while (!ctx->completed) {
+        SleepConditionVariableCS (&ctx->cv, &ctx->lock, INFINITE);
+    }
+    
+    LeaveCriticalSection(&ctx->lock);
+}
+
+void completion_context_init(completion_context_t *ctx) {
+    InitializeCriticalSection(&ctx->lock);
+    InitializeConditionVariable(&ctx->cv);
+}
+
+void completion_context_destroy(completion_context_t *ctx) {
+    DeleteCriticalSection(&ctx->lock);
+}
+
+long long get_time_ms(void) {
+    return GetTickCount();
```

Comment:
Might be better to at least use `GetTickCount64` to support systems with an uptime longer than 50 days.